std/assembly/string.ts

```diff
@@ -79,12 +79,13 @@ import { Array } from "./array";
     return (first - 0xD800 << 10) + (second - 0xDC00) + 0x10000;
   }
 
-  @operator("+") private static __concat(left: String, right: String): String {
-    return select<String>(left, changetype<String>("null"), left !== null).concat(right);
+  @operator("+") private static __concat(left: String | null, right: String | null): String {
+    if (right === null) return select<String>(left!, changetype<String>(""), left !== null);
+    if (left === null) return right;
+    return left.concat(right);
```

Comment:
So this guarantee `a + b` will always produce non-nullabe string.

```ts
let a: string | null = 'a';
let b: string | null = 'b';
a + b // 'ab'

b = null;
a + b // 'a'

a = null;
b = 'b';
a + b // 'b'

a = null;
b = null;
a + b; // '' 
```

Reply:
Another option is making this more stricter and accept only non-nullable strings for unary/binary string operations. In this case we can't write following anymore:
```ts
let a: string | null = 'a';
let b: string | null = 'b';
let c = a + b; // AS Error
```
and will use something like this:
```ts
let a: string | null = 'a';
let b: string | null = 'b';
let c = a! + b!;
```

Reply:
Yeah, that's the two options we have. My intuition currently is that if we allow `a + b` with one potentially `null`, we should deal with that like JS does, i.e. concatenating with the string `"null"`. It's fine to only do it in the operator overload, though, where it can happen, but we can remove the check everywhere else where the type is non-nullable anywhere, as mentioned in the previous comments.

Reply:
> My intuition currently is that if we allow a + b with one potentially null, we should deal with that like JS does, i.e. concatenating with the string "null"

I agree we should deal with nullish strings. But not sure about JS-way which convert null to "null" string. This is contrary to the general philosophy of AssemblyScript so it seems to me =) So I suggest have one special case for `<string>null + <string>null == ""` which looks more unexpected than "nullnull". See https://github.com/AssemblyScript/assemblyscript/pull/1682#discussion_r575676957

Reply:
I'm torn on this tbh. On the one hand what you suggest makes sense because concatenating with `null` doesn't really make a lot of sense anyhow, but on the other that'd be another deviation from JS behavior, and seems premature to do just because why not. I somewhat tend to not risk this to become a problem, or another peculiarity in the docs, and keep JS semantics in tact fwiw, that is unless we have a really compelling reason to change it.

Reply:
Perhaps an intuitive way to think of these things is in terms of "your fix may be someone else's bug" :)

Reply:
I thought this good reason (especially due to we support wasi):
```ts
const filename: string | null = null;
const path = "/dev/" + filename;
// > "/dev/null"
```

Reply:
> Perhaps an intuitive way to think of these things is in terms of "your fix may be someone else's bug" :)

C# support this:
```c#
string bob = "abc" + null + null + null + "123";  // abc123
```

Reply:
Hmmhmhm. I guess we may benefit from collecting a few opinions about this, and see what others would prefer in situations like these. Would you be interested in bringing this up in the next meeting, elaborate a bit on the what, how and why? :)

Reply:
Yeah, sure! Let's discuss it at the next meeting